internal/keppel/validation.go

```diff
@@ -6,37 +6,111 @@ package keppel
+		policy.RequiredLabels = generateRequiredLabels(account.RuleForManifest)
 	}
+
+	return &policy
 }
 
 // ApplyToAccount validates this policy and stores it in the given account model.
 func (v ValidationPolicy) ApplyToAccount(account *models.Account) *RegistryV2Error {
-	for _, label := range v.RequiredLabels {
-		if strings.Contains(label, ",") {
-			err := fmt.Errorf(`invalid label name: %q`, label)
+	// for backwards compatibility, if both validation.rule_for_manifest and validation.required_labels given,
+	// accept if and only if the presented CEL expression is the same as what we would generate from the presented list of required labels
+	if v.RuleForManifest != "" && v.RequiredLabels != nil {
+		generatedRule := generateRuleForManifest(v.RequiredLabels)
+		if generatedRule != v.RuleForManifest {
+			err := fmt.Errorf(`required labels %q do not match rule for manifest %q`, v.RequiredLabels, v.RuleForManifest)
 			return AsRegistryV2Error(err).WithStatus(http.StatusUnprocessableEntity)
 		}
 	}
+	if v.RuleForManifest != "" {
+		_, ast, celErr := BuildManifestValidationAST(v.RuleForManifest)
+		if celErr != nil {
+			err := fmt.Errorf(`invalid CEL expression: %q`, v.RuleForManifest)
+			return AsRegistryV2Error(err).WithStatus(http.StatusUnprocessableEntity).WithDetail(celErr.Error())
+		}
+		if ast.OutputType() != cel.BoolType {
+			err := fmt.Errorf(`output of CEL expression must be bool but is %q`, ast.OutputType().TypeName())
+			return AsRegistryV2Error(err).WithStatus(http.StatusUnprocessableEntity)
+		}
+		account.RuleForManifest = v.RuleForManifest
+	} else if v.RequiredLabels != nil {
+		// for backwards compatibility, if only validation.required_labels given,
+		// translate into a CEL expression
+		for _, label := range v.RequiredLabels {
+			if strings.Contains(label, ",") {
+				err := fmt.Errorf(`invalid label name: %q`, label)
+				return AsRegistryV2Error(err).WithStatus(http.StatusUnprocessableEntity)
+			}
+		}
+		account.RuleForManifest = generateRuleForManifest(v.RequiredLabels)
+	}
 
-	account.RequiredLabels = strings.Join(v.RequiredLabels, ",")
 	return nil
 }
+
+var celLabelExtractionRx = regexp.MustCompile(`has\(labels\.([a-zA-Z_][a-zA-Z0-9_]*)\)`)
+
+func generateRequiredLabels(ruleForManifest string) []string {
+	matches := celLabelExtractionRx.FindAllStringSubmatch(ruleForManifest, -1)
+
+	var labels []string
+	for _, match := range matches {
+		if len(match) > 1 {
+			labels = append(labels, match[1])
+		}
+	}
+
+	return labels
+}
+
+func generateRuleForManifest(requiredLabels []string) string {
+	var propositions []string
+	for _, label := range requiredLabels {
+		propositions = append(propositions, fmt.Sprintf("has(labels.%s)", label))
+	}
+
+	return strings.Join(propositions, " && ")
+}
+
+func BuildManifestValidationAST(celExpression string) (*cel.Env, *cel.Ast, error) {
+	env, err := cel.NewEnv(cel.Variable("labels", cel.MapType(cel.StringType, cel.StringType)))
```

Comment:
If `*cel.Env` is thread-safe (i.e. it is safe for us to use the same instance across multiple goroutines), then this can be a global variable that's instantiated once (with the error handling done through `must.Return()`).

The documentation does not explicitly state thread-safety, but [the implementation](https://github.com/google/cel-go/blob/v0.26.0/cel/env.go#L131) holds mutexes, so it appears to be designed that way.